### PR TITLE
Remove Mercury plugin dependency on raw evm chain.

### PR DESCRIPTION
**Pull request by brunotm**

### CODE REVIEW

File: go.mod
Comment (krehermann) [inline go.mod:390]: another way to do this simply `go get github.com/smartcontractkit/chainlink-relay@your_pr_commit` 

i find that easier than replacing. certainly, we must remove this replace and do go get before merging this

Comment (brunotm) [reply go.mod:390]: Had as a replacement while waiting for the related PR to be merged. Will drop that before merging.

File: core/services/relay/evm/mercury_provider.go
Comment (jmank88) [inline core/services/relay/evm/mercury_provider.go:137]: Let's store the sugared logger in lggr, instead of creating it every time.

**Review by jmank88 — DISMISSED** [review]

File: core/services/relay/evm/mercury_provider.go
Comment (brunotm) [reply core/services/relay/evm/mercury_provider.go]: dropped the logger since the previous PR for multiple block observation removed the fallback to the client as unnecessary.

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (jmank88) [inline core/services/relay/evm/mercury/v1/data_source.go:294]:
```suggestion
```

Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go]: thanks!

**Review by jmank88 — APPROVED** [review]

File: core/services/relay/evm/mercury_provider.go
Comment (brunotm) [reply core/services/relay/evm/mercury_provider.go]: I could not panic as an error is returned from LatestBlocks when no blocks are available.

Comment (brunotm) [reply core/services/relay/evm/mercury_provider.go]: I did drop the Head as it is not used anywhere.

File: core/services/relay/evm/mercury_provider.go
Comment (jmank88) [inline core/services/relay/evm/mercury_provider.go:131]: Hmm explicit error here seems like it would make things difficult for the caller. Why don't we leave it up to them to decided if this is an error or not?
```suggestion
		return nil, nil
```

Comment (jmank88) [reply core/services/relay/evm/mercury_provider.go:131]: Or we could generalize to checking length against `k`, and then reconsider this condition in that context.

Comment (brunotm) [reply core/services/relay/evm/mercury_provider.go]: Agreed. thanks!

**Comment by brunotm** [comment]
The current CI failure seems unrelated with this changeset.

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (jmank88) [inline core/services/relay/evm/mercury/v1/data_source.go:294]: Should this error be checked?

Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:294]: The reader on an evm chain always return a nil error, related with the previous review about letting the caller decide if an empty response is an error or not.

https://github.com/smartcontractkit/chainlink/pull/11201/files#diff-237ad2d55c5b2562977f53db7f2e1479118fde028622f031f86f9bf93c69c193R127

Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:294]: also the error here is not actually used. we can check for a non nil error on `len(latestBlocks) == 0` and use it as the observation error for consistency.

Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go]: Missed your last comment. Agreed, fixed to use the error on the observation.
CI is failing on a broken import related with the sqlx change.

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (jmank88) [inline core/services/relay/evm/mercury/v1/data_source.go:294]: Why don't we handle this error right away?

File: core/services/relay/evm/mercury_provider.go
Comment (samsondav) [inline core/services/relay/evm/mercury_provider.go:127]: This never returns an error

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:294]: The error is only used in the observation, what's your suggestion?

Comment (jmank88) [reply core/services/relay/evm/mercury/v1/data_source.go]: I'm not sure. It is unusual to interact with the value before checking the error. Will it be confusing that the logs for successful empty response look the same as a failed call? That seems significant.

File: core/services/relay/evm/mercury_provider.go
Comment (jmank88) [reply core/services/relay/evm/mercury_provider.go]: Every method in the relay API must return an error

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:293]: the interaction is the length check. I agree that it looks unusual, but the error not handled here but passed with the observation. @samsondav wdyt ?

Comment (jmank88) [reply core/services/relay/evm/mercury/v1/data_source.go:293]: Why exclude the error from the log line and drop it all together though? It won't be clear when there was a successful empty response vs an actual failure.

Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:293]: good point, will log.

**Review by jmank88 — DISMISSED** [review]

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (samsondav) [inline core/services/relay/evm/mercury/v1/data_source.go:295]: Might choose to panic here instead since we never expect this to happen

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (jmank88) [reply core/services/relay/evm/mercury/v1/data_source.go:295]: Sometimes a plugin will crash or be unavailable. This is expected to happen during "normal" operation and must be handled gracefully.

Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:295]: We can bubble it up if we think this should result in a pipeline execution error, but this is a change in the current behaviour of error handling and might deserve more attention. If so, we could revert the related changes in the chainlink-relay to unblock the build of core images.

Comment (samsondav) [reply core/services/relay/evm/mercury/v1/data_source.go:295]: In that case, can we make this method return error and error out in the calling function also?

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (brunotm) [reply core/services/relay/evm/mercury/v1/data_source.go:295]: by that you mean to Observe() to error out on  setLatestBlocks() error ?

File: core/services/relay/evm/mercury/v1/data_source.go
Comment (samsondav) [inline core/services/relay/evm/mercury/v1/data_source.go:299]: Probably this log line is redundant since the Observe error will be logged further up the stack but not a big deal since its so rare

**Review by samsondav — APPROVED** [review]

**Review by samsondav — APPROVED** [review]

File: core/services/relay/evm/mercury_provider.go
Comment (samsondav) [inline core/services/relay/evm/mercury_provider.go:137]:
```suggestion
			Hash:      evmBlocks[x].Hash.Bytes(),
```

**Review by samsondav — CHANGES_REQUESTED** [review]

**Review by samsondav — APPROVED** [review]

**Pull request merged**
